File: src/prod/src/data/txnreplicator/common/AtomicOperation.cpp
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

#include "stdafx.h"

using namespace ktl;
using namespace TxnReplicator;
using namespace Data::Utilities;

AtomicOperation::AtomicOperation(
    __in LONG64 transactionId,
    __in bool isPrimaryTransaction)
    : TransactionBase(transactionId, isPrimaryTransaction)
{
}

AtomicOperation::AtomicOperation(
    __in ITransactionManager & replicator,
    __in LONG64 transactionId,
    __in bool isPrimaryTransaction)
    : TransactionBase(replicator, transactionId, isPrimaryTransaction)
{
}

AtomicOperation::~AtomicOperation()
{
}

AtomicOperation::SPtr AtomicOperation::CreateAtomicOperation(
    __in ITransactionManager & replicator,
    __in KAllocator & allocator)
{
    LONG64 id = TransactionBase::CreateTransactionId();

    bool isPrimaryTx = replicator.Role == FABRIC_REPLICA_ROLE_PRIMARY ?
        true :
        false;

    AtomicOperation* pointer = _new(TRANSACTION_TAG, allocator) AtomicOperation(
        replicator, 
        -id, 
        isPrimaryTx);

    THROW_ON_ALLOCATION_FAILURE(pointer);
    return AtomicOperation::SPtr(pointer);
}

AtomicOperation::SPtr AtomicOperation::CreateAtomicOperation(
    __in LONG64 transactionId,
    __in bool isPrimaryTransaction,
    __in KAllocator & allocator)
{
    AtomicOperation* pointer = _new(TRANSACTION_TAG, allocator) AtomicOperation(
        transactionId,
        isPrimaryTransaction);

    THROW_ON_ALLOCATION_FAILURE(pointer);
    return AtomicOperation::SPtr(pointer);
}

Awaitable<NTSTATUS> AtomicOperation::AddOperationAsync(
    __in_opt OperationData const * const metaData,
    __in_opt OperationData const * const undo,
    __in_opt OperationData const * const redo,
    __in LONG64 stateProviderId,
    __in_opt OperationContext const * const operationContext) noexcept
{
    NTSTATUS status = ErrorIfNotPrimaryTransaction(IsPrimaryTransaction);

    if (!NT_SUCCESS(status))
    {
        co_return status;
    }

    ITransactionManager::SPtr txMgr;
    status = GetTransactionManager(txMgr);

    if (!NT_SUCCESS(status))
    {
        co_return status;
    }

    status = StateMachine->OnBeginAtomicOperationAddAsync();

    if (!NT_SUCCESS(status))
    {
        co_return status;
    }

    LONG64 lsn;
    status = co_await txMgr->AddOperationAsync(
        *this,
        metaData,
        undo,
        redo,
        stateProviderId,
        operationContext,
        lsn);

    if (NT_SUCCESS(status))
    {
        StateMachine->OnAtomicOperationSuccess();
        co_return status;
    }

    if (IsRetriableNTSTATUS(status))
    {
        StateMachine->OnAtomicOperationRetry();
    }
    else
    {
        StateMachine->OnAtomicOperationFaulted();
    }

    co_return status;
}

File: src/prod/src/managed/Microsoft.ServiceFabric.Data.Impl/Replicator/AtomicOperation.cs
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

namespace Microsoft.ServiceFabric.Replicator
{
    using System;
    using System.Fabric;
    using System.Threading.Tasks;

    /// <summary>
    /// An abstraction for a single operation transaction
    /// </summary>
    public class AtomicOperation : TransactionBase
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="AtomicOperation"/> class.
        /// </summary>
        /// <param name="transactionId">
        /// The transaction id.
        /// </param>
        /// <param name="transactionalReplicator">
        /// The transactional replicator.
        /// </param>
        /// <param name="isPrimaryTransaction">
        /// The is primary transaction.
        /// </param>
        private AtomicOperation(
            long transactionId,
            ITransactionalReplicator transactionalReplicator,
            bool isPrimaryTransaction)
            : base(transactionId, transactionalReplicator, isPrimaryTransaction)
        {
        }

        /// <summary>
        /// Adds an operation to the state of the atomic operation, that is immediately committed
        /// One atomic operation can contain only 1 operation:- and hence the name.
        /// Atomic operations cannot be 'committed' or 'rolled' back explicitly.
        /// </summary>
        /// <param name="metaData">metadata bytes pertaining to the operation.</param>
        /// <param name="undo">Byte stream that contains the 'undo' data when this atomic operation is to be aborted and the effects of this operation are to be undone</param>
        /// <param name="redo">Byte stream that contains the 'redo' data when this atomic operation is to be committed and the effects of this operation are to be applied</param>
        /// <param name="operationContext"> A user defined context</param>
        /// <param name="stateProviderId">Id of the state provider</param>
        /// <exception cref="FabricNotPrimaryException">
        /// If replica role is not primary
        /// </exception>
        /// <exception cref="InvalidOperationException">
        /// If Transaction is not active (Already committed or aborted)
        /// </exception>
        public async Task AddOperationAsync(
            OperationData metaData, 
            OperationData undo, 
            OperationData redo, 
            object operationContext, 
            long stateProviderId)
        {
            // Can throw InvalidOperationException.
            // Note that this check protects against an Assert in SM that checks that Apply is only called with transactions that have
            // isPrimaryTransaction set to true.
            this.ThrowIfNotPrimaryTransaction();
            this.ThrowIfNotPrimary();

            StateMachine.OnBeginAtomicOperationAddAsync();

            try
            {
                await TransactionalReplicator.AddOperationAsync(this, metaData, undo, redo, operationContext, stateProviderId).ConfigureAwait(false);
                StateMachine.OnAtomicOperationSuccess();
            }
            catch (Exception e)
            {
                var isRetryable = ProcessException(e);

                // If this is a retryable exception, set the this.TransactionalReplicator so that the operation can be retried by the user
                if (isRetryable)
                {
                    StateMachine.OnAtomicOperationRetry();
                    throw;
                }

                StateMachine.OnAtomicOperationFaulted();
                throw;
            }
        }

        /// <summary>
        /// The create atomic operation.
        /// </summary>
        /// <param name="transactionId">
        /// The transaction id.
        /// </param>
        /// <param name="isPrimaryTransaction">
        /// The is primary transaction.
        /// </param>
        /// <returns>
        /// The <see cref="AtomicOperation"/>.
        /// </returns>
        internal static AtomicOperation CreateAtomicOperation(long transactionId, bool isPrimaryTransaction)
        {
            Utility.Assert(transactionId < 0, "transactionId < 0");
            return new AtomicOperation(transactionId, null, isPrimaryTransaction);
        }

        /// <summary>
        /// Creates an instance of an atomic operation
        /// </summary>
        /// <param name="transactionalReplicator"> Transactional Replicator object</param>
        /// <returns> An atomic operation</returns>
        internal static AtomicOperation CreateAtomicOperationInternal(ITransactionalReplicator transactionalReplicator)
        {
            Utility.Assert(
                transactionalReplicator != null,
                "CreateAtomicOperationInternal: transactionalReplicator is null");

            var isPrimaryTransaction = transactionalReplicator.Role == ReplicaRole.Primary;

            var transactionId = CreateTransactionId();
            return new AtomicOperation(-transactionId, transactionalReplicator, isPrimaryTransaction);
        }
    }
}